examples/fabric-admin/device_manager/UidGetter.cpp

```diff
@@ -0,0 +1,154 @@
+                             chip::EndpointId endpointId)
+{
+    assertChipStackLockedByCurrentThread();
+    VerifyOrDie(!mCurrentlyGettingUid);
+
+    mEndpointId       = endpointId;
+    mOnDoneCallback   = onDoneCallback;
+    mUniqueIdHasValue = false;
+    memset(mUniqueId, 0, sizeof(mUniqueId));
+    mCurrentlyGettingUid = true;
+
+    CHIP_ERROR err = controller.GetConnectedDevice(nodeId, &mOnDeviceConnectedCallback, &mOnDeviceConnectionFailureCallback);
+    if (err != CHIP_NO_ERROR)
+    {
+        mCurrentlyGettingUid = false;
+    }
+    return err;
+}
+
+void UidGetter::OnAttributeData(const ConcreteDataAttributePath & path, TLV::TLVReader * data, const StatusIB & status)
+{
+    VerifyOrDie(path.mClusterId == Clusters::BridgedDeviceBasicInformation::Id);
+
+    if (!status.IsSuccess())
+    {
+        ChipLogError(NotSpecified, "Response Failure: %" CHIP_ERROR_FORMAT, status.ToChipError().Format());
+        return;
+    }
+
+    switch (path.mAttributeId)
+    {
+    case Clusters::BridgedDeviceBasicInformation::Attributes::UniqueID::Id: {
+        mUniqueIdHasValue = SuccessOrLog(data->GetString(mUniqueId, sizeof(mUniqueId)), "UniqueId");
+        break;
+    }
+    default:
+        break;
+    }
+}
+
+void UidGetter::OnReportEnd()
+{
+    // We will call mOnDoneCallback in OnDone.
+}
+
+void UidGetter::OnError(CHIP_ERROR error)
+{
+    ChipLogProgress(NotSpecified, "Error Getting UID: %" CHIP_ERROR_FORMAT, error.Format());
+}
+
+void UidGetter::OnDone(ReadClient * apReadClient)
+{
+    mCurrentlyGettingUid = false;
+    std::optional<CharSpan> returnValue;
+    if (mUniqueIdHasValue)
+    {
+        returnValue = CharSpan(mUniqueId);
+    }
+    mOnDoneCallback(returnValue);
+}
+
+void UidGetter::OnDeviceConnected(Messaging::ExchangeManager & exchangeMgr, const SessionHandle & sessionHandle)
+{
+    VerifyOrDie(mCurrentlyGettingUid);
+    mClient = std::make_unique<ReadClient>(app::InteractionModelEngine::GetInstance(), &exchangeMgr, *this /* callback */,
+                                           ReadClient::InteractionType::Read);
+    VerifyOrDie(mClient);
+
+    AttributePathParams readPaths[1];
+    readPaths[0] = AttributePathParams(mEndpointId, Clusters::BridgedDeviceBasicInformation::Id,
+                                       Clusters::BridgedDeviceBasicInformation::Attributes::UniqueID::Id);
+
+    ReadPrepareParams readParams(sessionHandle);
+
+    readParams.mpAttributePathParamsList    = readPaths;
+    readParams.mAttributePathParamsListSize = 1;
+
+    CHIP_ERROR err = mClient->SendRequest(readParams);
+
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(NotSpecified, "Failed to issue subscription to AdministratorCommissioning data");
+        // After calling mOnDoneCallback we are indicating that `this` is deleted and we shouldn't do anything else with
+        // DeviceSubscription.
+        mCurrentlyGettingUid = false;
+        mOnDoneCallback(std::nullopt);
```

Comment:
Did you consider calling `OnDone(nullptr)` here? That we we would have a single place changing mCurrentlyGettingUid and mOnDoneCallback, however the downside of having OnDone being passed in a nullptr potentially. We could even do `OnDone(mClient.get())` if we want non-null arguments to it.